pyro/contrib/timeseries/gp.py

```diff
@@ -422,3 +426,58 @@ def forecast(self, targets, dts):
         filtering_state = self._filter(targets)
         predicted_mean, predicted_covar = self._forecast(dts, filtering_state)
         return MultivariateNormal(predicted_mean, predicted_covar)
+
+
+class LinearlyCoupledDependentMaternGP(DependentMaternGP):
+    """
+    A time series model in which each output dimension is modeled as a linear combination
+    of univariate Gaussian Processes with Matern kernels. The different output dimensions
+    are additionally correlated because the Gaussian Processes are driven by a correlated
+    Wiener process. That is, this time series model combines the structure of the
+    `LinearlyCoupledMaternGP` (with `num_gps` equal to `obs_dim`) with that of the
+    `DependentMaternGP`. The targets are assumed to be evenly spaced in time. Training
+    and inference are logarithmic in the length of the time series T.
+
+    :param float nu: The order of the Matern kernel; must be 1.5.
+    :param float dt: The time spacing between neighboring observations of the time series.
+    :param int obs_dim: The dimension of the targets at each time step.
+    :param torch.Tensor log_length_scale_init: optional initial values for the kernel length scale
+        given as a `obs_dim`-dimensional tensor
+    :param torch.Tensor log_obs_noise_scale_init: optional initial values for the observation noise scale
+        given as a `obs_dim`-dimensional tensor
+    """
+    def __init__(self, nu=1.5, dt=1.0, obs_dim=1,
+                 log_length_scale_init=None, log_obs_noise_scale_init=None):
+
+        if nu != 1.5:
+            raise NotImplementedError("The only supported value of nu is 1.5")
+
+        self.dt = dt
+        self.obs_dim = obs_dim
+
+        if log_obs_noise_scale_init is None:
+            log_obs_noise_scale_init = -2.0 * torch.ones(obs_dim)
+        assert log_obs_noise_scale_init.shape == (obs_dim,)
+
+        nn.Module.__init__(self)
+
+        self.kernel = MaternKernel(nu=nu, num_gps=obs_dim,
+                                   log_length_scale_init=log_length_scale_init)
+        self.full_state_dim = self.kernel.state_dim * obs_dim
+
+        # we demote self.kernel.log_kernel_scale from being a nn.Parameter
+        # since the relevant scales are now encoded in the wiener noise matrix
+        del self.kernel.log_kernel_scale
+        self.kernel.register_buffer("log_kernel_scale", torch.zeros(obs_dim))
+
+        self.log_obs_noise_scale = nn.Parameter(log_obs_noise_scale_init)
+        self.log_diag_wiener_noise = nn.Parameter(torch.zeros(obs_dim))
+        self.off_diag_wiener_noise = nn.Parameter(0.03 * torch.randn(obs_dim, obs_dim))
+
+        self.obs_matrix = nn.Parameter(0.3 * torch.randn(self.obs_dim, self.obs_dim))
+
+    def _get_obs_matrix(self):
```

Comment:
basically the only thing different about this model is that the obs_matrix is learned (instead of just being a particular matrix made of 0s and 1s)

Reply:
Thanks for explaining.